cpp/src/packed/chunk_manager.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright 2024 Zilliz
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <arrow/array/array_base.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+#include <parquet/properties.h>
+#include "packed/chunk_manager.h"
+
+#include <iostream>
+
+using namespace std;
+
+namespace milvus_storage {
+
+ChunkManager::ChunkManager(const std::vector<ColumnOffset>& column_offsets) : column_offsets_(column_offsets) {
+  chunk_states_ = std::vector<ChunkState>(column_offsets_.size());
+  for (int i = 0; i < column_offsets_.size(); ++i) {
+    chunk_states_[i] = ChunkState(0, 0);
+  }
+}
+
+std::vector<std::shared_ptr<arrow::ArrayData>> ChunkManager::ProcessChunks(
+    const std::vector<std::shared_ptr<arrow::Table>>& tables, int64_t chunksize) {
+  std::vector<const arrow::Array*> chunks = CollectChunks(tables, chunksize);
+
+  // Slice chunks and advance chunk index as appropriate
+  std::vector<std::shared_ptr<arrow::ArrayData>> batch_data(column_offsets_.size());
+  for (int i = 0; i < column_offsets_.size(); ++i) {
+    batch_data[i] = SliceChunk(chunks[i], i, chunksize);
+  }
+  return batch_data;
+}
+
+std::vector<const arrow::Array*> ChunkManager::CollectChunks(const std::vector<std::shared_ptr<arrow::Table>>& tables,
```

Comment:
Returning a vector of `std::shared_ptr<arrow::Array>` is probably better, or else we must handle the lifecycle of the pointers.